cpp/src/sampling/neighbor_sampling_impl.hpp

```diff
@@ -217,18 +261,70 @@ neighbor_sample_impl(raft::handle_t const& handle,
                                   starting_vertex_labels);
       }
 
-      level_sizes.push_back(srcs.size());
-      level_result_src_vectors.push_back(std::move(srcs));
-      level_result_dst_vectors.push_back(std::move(dsts));
+      auto old_size = level_result_src.size();
+      level_result_src.resize(old_size + srcs.size(), handle.get_stream());
+      level_result_dst.resize(old_size + srcs.size(), handle.get_stream());
 
-      if (weights) { (*level_result_weight_vectors).push_back(std::move(*weights)); }
-      if (edge_ids) { (*level_result_edge_id_vectors).push_back(std::move(*edge_ids)); }
-      if (edge_types) { (*level_result_edge_type_vectors).push_back(std::move(*edge_types)); }
-      if (labels) { (*level_result_label_vectors).push_back(std::move(*labels)); }
+      raft::copy(
+        level_result_src.begin() + old_size, srcs.begin(), srcs.size(), handle.get_stream());
+
+      raft::copy(
+        level_result_dst.begin() + old_size, dsts.begin(), srcs.size(), handle.get_stream());
+
+      if (weights) {
+        (*level_result_weight).resize(old_size + srcs.size(), handle.get_stream());
+
+        raft::copy(level_result_weight->begin() + old_size,
+                   weights->begin(),
+                   srcs.size(),
+                   handle.get_stream());
+      }
+
+      if (edge_ids) {
+        (*level_result_edge_id).resize(old_size + srcs.size(), handle.get_stream());
+        raft::copy(level_result_edge_id->begin() + old_size,
+                   edge_ids->begin(),
+                   srcs.size(),
+                   handle.get_stream());
+      }
+      if (edge_types) {
+        (*level_result_edge_type).resize(old_size + srcs.size(), handle.get_stream());
+
+        raft::copy(level_result_edge_type->begin() + old_size,
+                   edge_types->begin(),
+                   srcs.size(),
+                   handle.get_stream());
+      }
+
+      if (labels) {
+        (*level_result_label).resize(old_size + srcs.size(), handle.get_stream());
+
+        raft::copy(level_result_label->begin() + old_size,
+                   labels->begin(),
+                   srcs.size(),
+                   handle.get_stream());
+      }
 
       if (num_edge_types > 1) { modified_graph_view.clear_edge_mask(); }
     }
 
+    level_sizes.push_back(level_result_src.size());
+    level_result_src_vectors.push_back(std::move(level_result_src));
+    level_result_dst_vectors.push_back(std::move(level_result_dst));
```

Comment:
After calling std::move, the variable is in valid but arbitrary state. You should not use re-use the moved-out variable assuming that it's size is reset to 0. And to minimize the scope of per-level (per-hop) variables, we should better define these variables inside the `for (auto hop = 0; hop < num_hops; hop++) { ... }`.

Reply:
> The input vertex labels are const and won't be modified in this function (even though you're updating starting_vertices to point different arrays).

As you mentioned, vertex labels are constant hence it can't be sorted and I need the initial labels downstream. I reduced the scope of its copy